core/src/main/java/org/commonjava/maven/ext/manip/impl/GroovyManipulator.java

```diff
@@ -0,0 +1,146 @@
+import org.slf4j.LoggerFactory;
+
+import java.io.File;
+import java.io.IOException;
+import java.util.Collections;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Properties;
+import java.util.Set;
+
+/**
+ * {@link Manipulator} implementation that can resolve a remote groovy file and execute it on executionRoot. Configuration
+ * is stored in a {@link org.commonjava.maven.ext.manip.state.GroovyState} instance, which is in turn stored in the {@link ManipulationSession}.
+ */
+@Component( role = Manipulator.class, hint = "groovy-injection" )
+public class GroovyManipulator
+    implements Manipulator
+{
+    private final Logger logger = LoggerFactory.getLogger( getClass() );
+
+    @Requirement
+    protected ModelIO modelBuilder;
+
+    /**
+     * No prescanning required for Profile injection.
+     */
+    @Override
+    public void scan( final List<Project> projects, final ManipulationSession session )
+        throws ManipulationException
+    {
+    }
+
+    /**
+     * Initialize the {@link GroovyState} state holder in the {@link ManipulationSession}. This state holder detects
+     * version-change configuration from the Maven user properties (-D properties from the CLI) and makes it available for
+     * later invocations of {@link GroovyManipulator#scan(List, ManipulationSession)} and the apply* methods.
+     */
+    @Override
+    public void init( final ManipulationSession session )
+    {
+        final Properties userProps = session.getUserProperties();
+        session.setState( new GroovyState( userProps ) );
+    }
+
+    /**
+     * Apply the groovy script changes to the top level pom.
+     */
+    @Override
+    public Set<Project> applyChanges( final List<Project> projects, final ManipulationSession session )
+        throws ManipulationException
+    {
+        final GroovyState state = session.getState( GroovyState.class );
+        if ( !session.isEnabled() || !state.isEnabled() )
+        {
+            logger.debug( getClass().getSimpleName() + ": Nothing to do!" );
+            return Collections.emptySet();
+        }
+
+        final Set<Project> changed = new HashSet<Project>();
+        final List<ArtifactRef> scripts = state.getGroovyScripts();
+
+        for (ArtifactRef ar : scripts)
+        {
+            logger.info ("Attempting to read GAV {} with classifier {} and type {} ",
+                         ar.asProjectVersionRef(), ar.getClassifier(), ar.getType());
+
+            final File groovyScript = modelBuilder.resolveRawFile( ar );
+
+            Binding binding = new Binding( );
+            GroovyShell shell = new GroovyShell( binding );
+            Script script = null;
+
+            for ( final Project project : projects )
+            {
+                if ( project.isExecutionRoot() )
+                {
+                    binding.setProperty( "basedir", project.getPom().getParentFile().toString() );
+                    binding.setProperty( "name", project.getKey() );
+
```

Comment:
I'd probably bind the projects too, just to avoid any confusion in rewriting them once in groovy, then again in java...if that happened, the java rewrite would erase the changes from the groovy script, since the project in memory was read ahead of time.